pkg/osbuild/mkfs_stage.go

```diff
@@ -7,81 +7,96 @@ import (
 	"github.com/osbuild/images/pkg/disk"
 )
 
-// GenMkfsStages generates a list of org.mkfs.* stages based on a
-// partition table description for a single device node
-// filename is the path to the underlying image file (to be used as a source for the loopback device)
-func GenMkfsStages(pt *disk.PartitionTable, filename string) []*Stage {
+// GenFsStages generates a list of stages that create the filesystem and other
+// related entities. Specifically, it creates stages for:
+//   - org.osbuild.mkfs.*: for all filesystems and btrfs volumes
+//   - org.osbuild.btrfs.subvol: for all btrfs subvolumes
+//   - org.osbuild.mkswap: for swap areas
+func GenFsStages(pt *disk.PartitionTable, filename string) []*Stage {
 	stages := make([]*Stage, 0, len(pt.Partitions))
 
-	processedBtrfsPartitions := make(map[string]bool)
-	genStage := func(mnt disk.Mountable, path []disk.Entity) error {
-		t := mnt.GetFSType()
-		var stage *Stage
+	genStage := func(ent disk.Entity, path []disk.Entity) error {
+		switch e := ent.(type) {
+		case *disk.Filesystem:
+			// TODO: extract last device renaming into helper
```

Comment:
Fwiw, I did a quick experiemnt on this in https://github.com/osbuild/images/commit/fa11558c23af57532a0b41d5f676726393a21843 - I would love to see this slightly cleaner (the linked patch is still not nice but avoids some of the repetition and is hopefully a starting point)

Reply:
I left this as a TODO because I didn't want to add extra things and it could use some investigation.  Sometimes we need to rename the leaf device and other times we don't, but maybe renaming it always is a good general approach and we can just add it tot he existing function.  Alternatively we can have a "rename last device" function.  Another alternative is to add an argument to the function to rename the last device to whatever the caller specifies.

I don't know which is cleaner, so I'd rather look into it separately :)